pkg/verifier/codeowner.go

```diff
@@ -0,0 +1,213 @@
+
+	log "github.com/sirupsen/logrus"
+	"github.com/topfreegames/codeowners-verifier/pkg/providers"
+)
+
+// CodeOwner represents a line in a CODEOWNERS file
+type CodeOwner struct {
+	Path   string
+	Regex  *regexp.Regexp
+	Line   int
+	Owners []string
+	Negate bool
+}
+
+// reverseCodeOwners returns an inverted slice
+func reverseCodeOwners(a []*CodeOwner) []*CodeOwner {
+	for left, right := 0, len(a)-1; left < right; left, right = left+1, right-1 {
+		a[left], a[right] = a[right], a[left]
+	}
+	return a
+}
+
+// Used to remove comment lines
+const commentChars = "[#]"
+
+// stripComment uses the commentChars to remove comments from lines
+func stripComment(source string) string {
+	if cut := strings.IndexAny(source, commentChars); cut >= 0 {
+		return strings.TrimRightFunc(source[:cut], unicode.IsSpace)
+	}
+	return source
+}
+
+// difference returns the difference between two slices
+func difference(slice1 []string, slice2 []string) []string {
+	diffStr := []string{}
+
+	for _, s1Val := range slice1 {
+		found := false
+		for _, s2Val := range slice2 {
+			if s1Val == s2Val {
+				found = true
+				break
+			}
+		}
+		if !found {
+			diffStr = append(diffStr, s1Val)
+		}
+	}
+	return diffStr
+}
+
+// ReadCodeownersFile reads the file specified by filename
+// and returns a list of CodeOwners strucs, as well as an error
+func ReadCodeownersFile(filename string) ([]*CodeOwner, error) {
+	var codeowners []*CodeOwner
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't open file: %s", err)
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	lineNumber := 1
+	for scanner.Scan() {
+		line := stripComment(scanner.Text())
+		line = strings.Join(strings.Fields(line), " ")
+		if line != "" {
+			elements := strings.Split(line, " ")
+			if len(elements) < 2 {
+				return nil, fmt.Errorf("Invalid CODEOWNERS entry: %d", lineNumber)
+			}
+			regex, negateRegex := getPatternFromLine(elements[0])
+			if regex != nil {
+				c := &CodeOwner{
+					Path:   elements[0],
+					Regex:  regex,
+					Line:   lineNumber,
+					Owners: elements[1:],
+					Negate: negateRegex,
+				}
+				codeowners = append(codeowners, c)
+			}
+		}
+		lineNumber++
+	}
+	return codeowners, nil
+}
+
+// ValidateCodeownerFile check if every entry:
+// 1. Has a valid file/path
+// 2. Check if every owner is an user or a group.
+func ValidateCodeownerFile(p providers.Provider, filename string) (bool, error) {
+	codeowners, err := ReadCodeownersFile(filename)
+	if err != nil {
+		return false, err
+	}
+	valid := true
+	for _, c := range codeowners {
+		// This doesn't works for ! patterns
+		if files, _, err := glob.Glob([]string{c.Path}); err != nil || len(files) < 1 {
+			log.Errorf("Error parsing line %d, path %s does not exist", c.Line, c.Path)
+			valid = false
+		}
+		for _, element := range c.Owners {
+			owner := strings.Replace(element, "@", "", 1)
+			exists, err := p.UserExists(owner)
+			if err != nil || !exists {
```

Comment:
Can we differentiate the errors checks from the existence check?

It would improve the error log message giving a better feedback. Does the code failed to get the user/group or it just didn't find the user.